File: ios/OpenPatent/Core/Networking/AuthRepository.swift
import Foundation

protocol AuthRepositoryProtocol {
    func login(email: String, password: String) async throws -> AuthResponse
    func register(email: String, password: String, username: String?) async throws -> AuthResponse
    func refreshToken(_ refresh: String) async throws -> TokenResponse
    func logout() async throws
    func getCurrentUser() async throws -> UserResponse
}

final class AuthRepository: AuthRepositoryProtocol {
    private let api: ApiClientProtocol
    
    init(api: ApiClientProtocol) {
        self.api = api
    }
    
    func login(email: String, password: String) async throws -> AuthResponse {
        try await api.request(.login, body: LoginRequest(email: email, password: password))
    }
    
    func register(email: String, password: String, username: String?) async throws -> AuthResponse {
        try await api.request(.register, body: RegisterRequest(email: email, password: password, username: username))
    }
    
    func refreshToken(_ refresh: String) async throws -> TokenResponse {
        struct RefreshRequest: Encodable {
            let refresh: String
        }
        let response: TokenResponse = try await api.request(.refreshToken, body: RefreshRequest(refresh: refresh))
        return response
    }
    
    func logout() async throws {
        let _: EmptyResponse = try await api.request(.logout, body: nil)
    }
    
    func getCurrentUser() async throws -> UserResponse {
        try await api.request(.getMe, body: nil)
    }
}

private struct EmptyResponse: Decodable {}

File: android/app/src/main/java/ai/openpatent/data/repository/AuthRepository.kt
package ai.openpatent.data.repository

import ai.openpatent.data.local.SecureTokenStorage
import ai.openpatent.data.model.*
import ai.openpatent.data.remote.api.AuthApi
import kotlinx.coroutines.flow.Flow
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class AuthRepository @Inject constructor(
    private val authApi: AuthApi,
    private val tokenStorage: SecureTokenStorage
) {
    val isLoggedIn: Flow<Boolean> = tokenStorage.isLoggedIn
    
    suspend fun login(email: String, password: String): Result<User> {
        return try {
            val response = authApi.login(LoginRequest(email, password))
            tokenStorage.saveTokens(response.access, response.refresh)
            Result.success(response.user)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun register(email: String, password: String, username: String? = null): Result<User> {
        return try {
            val response = authApi.register(RegisterRequest(email, password, username))
            tokenStorage.saveTokens(response.access, response.refresh)
            Result.success(response.user)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun getCurrentUser(): Result<UserResponse> {
        return try {
            val user = authApi.getCurrentUser()
            Result.success(user)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun refreshToken(): Result<String> {
        return try {
            val refreshToken = tokenStorage.getRefreshTokenSync()
                ?: return Result.failure(Exception("No refresh token"))
            
            val response = authApi.refreshToken(RefreshRequest(refreshToken))
            tokenStorage.updateAccessToken(response.access)
            response.refresh?.let { tokenStorage.saveTokens(response.access, it) }
            
            Result.success(response.access)
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
    
    suspend fun logout(): Result<Unit> {
        return try {
            val refreshToken = tokenStorage.getRefreshTokenSync()
            if (refreshToken != null) {
                try {
                    authApi.logout(LogoutRequest(refreshToken))
                } catch (e: Exception) {
                    // Ignore logout API errors, still clear local tokens
                }
            }
            tokenStorage.clearTokens()
            Result.success(Unit)
        } catch (e: Exception) {
            tokenStorage.clearTokens()
            Result.success(Unit)
        }
    }
    
    fun isAuthenticated(): Boolean {
        return tokenStorage.getAccessTokenSync() != null
    }
}
